validator/sawtooth_validator/networking/dispatch.py

```diff
@@ -195,11 +202,13 @@ class _ManagerCollection(object):
     def __init__(self, handler_managers):
         self._chain = handler_managers
         self._index = 0
+        self._lock = Lock()
 
     def __next__(self):
-        result = self._chain[self._index]
-        self._index += 1
-        return result
+        with self._lock:
```

Comment:
It will throw KeyError, which I could catch and raise StopIteration, but that would only be needed in a for-loop. Its current use is having next() directly called on it, and catching the KeyError if it is thrown. I don't see the "the index will get advanced by each thread, which may result in them getting different iteration results" as a bad thing, as this data structure is used in the message_information dict to see which Handler has been executed on a message.